#7960 Add support for Markdown footnote syntax

tclaus opened the pull request:
It solves #7960 Add support for Markdown footnote syntax

A quick fix, but surely rarely used. 

It adds the possibility to add footnotes to posts. Syntax is defined here: https://pandoc.org/MANUAL.html#footnotes

Source example: 
![Bildschirmfoto 2021-07-31 um 08 01 17](https://user-images.githubusercontent.com/501326/127731790-a15a188c-09aa-40de-99be-6234c01f61f8.png)

The outcome: 
![Bildschirmfoto 2021-07-31 um 08 24 48](https://user-images.githubusercontent.com/501326/127731796-fc12db9a-1bb4-454e-b135-1c5b9b8eca8c.png)

It would be nice if this could then be added to the diaspora markdown - doc: 
https://diasporafoundation.org/formatting

tclaus committed:
#7960 Add support for Markdown footnote syntax

diaspora-code-review commented on Gemfile:
```diff
@@ -105,10 +105,11 @@ source "https://gems.diasporafoundation.org" do
-  gem "rails-assets-markdown-it-hashtag",                 "0.4.0"
   gem "rails-assets-markdown-it-diaspora-mention",        "1.2.0"
-  gem "rails-assets-markdown-it-sanitizer",               "0.4.3"
+  gem "rails-assets-markdown-it-footnote",                "3.0.3"
```
Bundler/OrderedGems: Gems should be sorted in an alphabetical order within their section of the Gemfile. Gem `rails-assets-markdown-it-diaspora-mention` should appear before `rails-assets-markdown-it-hashtag`.

tclaus replied:
It appears before ..-hashtag! Trouble with Gemfiles?

SuperTux88 replied:
The review-bot uses it's own Gemfile, so this can be ignored and probably also fixes that problem for the future

tclaus commented:
Nice.. https://societas.online/posts/81f78e60d40301392d8042010a9c0002

tclaus committed:
#7960 Add support for Markdown footnote syntax in Mobile view

Splitting stripdown and markdownify options, because the footnote option seems to let the Redcarpet::Render::StripDown function crash

SuperTux88 commented on lib/diaspora/message_renderer.rb:
```diff
@@ -50,7 +50,7 @@ def escape
 
       def strip_markdown
-        renderer = Redcarpet::Markdown.new Redcarpet::Render::StripDown, options[:markdown_options]
+        renderer = Redcarpet::Markdown.new Redcarpet::Render::StripDown, options[:stripdown_options]
```
Why does this need special `stripdown_options` without `footnotes`? So text-only doesn't support footnotes?

I think the reason for this should be commented here. And I'd also prefer still using `options[:markdown_options]` but just explicitly dropping `footnotes` from it (or set it to `false`?), so there is still only one `markdown_options` in `DEFAULTS` and if `markdown_options` needs to be changed it only needs to be changed once. Otherwise maybe in the future only `markdown_options` gets updated and `stripdown_options` is forgotten.

tclaus committed:
#7960 Using shared markdown options for plain text

SuperTux88 commented on lib/diaspora/message_renderer.rb:
```diff
@@ -50,7 +50,10 @@ def escape
-        renderer = Redcarpet::Markdown.new Redcarpet::Render::StripDown, options[:markdown_options]
+        stripdown_options = options[:markdown_options]
+        # Footnotes are not supported in text-only outputs (mail, crossposts etc)
+        stripdown_options[:footnotes] = false
```
This actually changes the value in `options` since it's only a reference to the original options. I don't know if it's a problem right now (if the same message renderer is uses twice), but I think it's at least an unexpected side-effect. I'd just do `stripdown_options = options[:markdown_options].except(:footnotes)` which gives you a new hash without that key, but without changing the original.

tclaus replied:
Thanks for the code-suggestion and explanation ;-)

tclaus committed:
Fix review issues

SuperTux88 commented:
Merged as 122e12a0689856f0da227603d0bfe04bb5a814a3 c71caa5cb8479686a1ca3cac7521f2047de3cb0b 8f9ac33649e4d711ad2d39cd8994fe9c824a9902

Thanks for the PR :cookie:
